README.rst
=======
pipstat
=======

DEPRECATED
==========

pipstat has been deprecated. Use the newer and actively maintained `pypi-stat`_

.. _pypi-stat: https://github.com/sloria/pypi-cli


Get download statistics for PyPI packages from the command line.
::

    $ pipstat marshmallow
    Fetching statistics for 'marshmallow'. . .

    Download statistics for marshmallow
    ===================================
    Downloads by version
    0.1.0  13/11/11 [ 818 ] *****************************************************
    0.2.0  13/11/12 [ 855 ] ********************************************************
    0.2.1  13/11/12 [ 806 ] ****************************************************
    0.3.0  13/11/14 [ 760 ] *************************************************
    0.3.1  13/11/16 [ 695 ] *********************************************
    0.4.0  13/11/25 [ 641 ] *****************************************
    0.4.1  13/12/02 [ 695 ] *********************************************
    0.5.0  13/12/29 [ 719 ] ***********************************************
    0.5.1  14/02/03 [ 316 ] ********************

    Min downloads:            316 (0.5.1)
    Max downloads:            855 (0.2.0)
    Avg downloads:            700
    Total downloads:        6,305

Get it now
----------
::

    pip install pipstat


Requirements
------------

- Python >= 2.7 or >= 3.3

License
-------

MIT licensed. See the bundled `LICENSE <https://github.com/sloria/pipstat/blob/master/LICENSE>`_ file for more details.


HISTORY.rst
*********
Changelog
*********


0.3.0 (unreleased)
******************

* Switched from XMLRPC API to JSON API, which is much faster and has more information.
* Can input full PyPI URLs.


test_pipstat.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from subprocess import check_output, CalledProcessError
import pytest
import random
from datetime import datetime

import pipstat

random.seed(42)

# Build up a dict of release numbers => download counts
MOCK_PACKAGE = 'cheesetest'
versions = ['0.1', '0.1.1', '0.1.12', '0.2.0']
release_downloads = {}
release_urls = {}
for i, ver in enumerate(versions):
    tar = '{0}-{1}.tar.gz'.format(MOCK_PACKAGE, ver)
    whl = '{0}-{1}-py2.py3-none-any.whl'.format(MOCK_PACKAGE, ver)
    tar_downloads =  random.randint(0, 12345)
    whl_downloads = random.randint(0, 12345)
    release_downloads[ver] = [[tar, tar_downloads], [whl, whl_downloads]]
    release_urls[ver] = [
        {'downloads': tar_downloads, 'upload_date': datetime(2014, 5, i + 1)},
        {'downloads': whl_downloads, 'upload_date': datetime(2014, 5, i + 1)}
    ]


class MockClient(object):
    """A simple mock for xmlrpclib.ServerProxy."""

    def package_releases(self, name, hidden=False):
        if name == MOCK_PACKAGE and hidden:
            return ['0.1', '0.1.1', '0.1.12', '0.2.0']
        else:
            return []

    def release_downloads(self, name, version):
        if name == MOCK_PACKAGE:
            return release_downloads[version]
        else:
            return []

    def release_urls(self, name, version):
        if name == MOCK_PACKAGE:
            return release_urls[version]
        else:
            return []


@pytest.fixture
def package():
    return pipstat.Package(MOCK_PACKAGE, client=MockClient())


def test_versions(package):
    assert package.versions == package.client.package_releases(MOCK_PACKAGE, True)


def test_error_raised_if_no_versions():
    p = pipstat.Package('notfound', client=MockClient())
    with pytest.raises(pipstat.NotFoundError):
        p.versions


def test_repr(package):
    assert repr(package) == 'Package(name={0!r})'.format(MOCK_PACKAGE)


def test_version_downloads(package):
    for ver in package.versions:
        expected = sum(dl for _, dl in release_downloads[ver])
        assert package.version_downloads[ver] == expected


def test_downloads(package):
    assert package.downloads == sum(package.version_downloads.values())


def test_max_version(package):
    assert package.max_version == max(package.version_downloads.items(),
                                        key=lambda item: item[1])


def test_min_version(package):
    assert package.min_version == min(package.version_downloads.items(),
                                        key=lambda item: item[1])


def test_avg_downloads(package):
    avg = package.downloads / len(release_urls.keys())
    assert package.average_downloads == int(avg)


def test_pipstat_cmd_without_args_exits_with_nonzero_code():
    with pytest.raises(CalledProcessError):
        run_cmd('pipstat')


def test_pipstat_cmd_version():
    assert run_cmd('pipstat -v') == pipstat.__version__ + '\n'
    assert run_cmd('pipstat --version') == pipstat.__version__ + '\n'


def run_cmd(cmd):
    '''Run a shell command `cmd` and return its output.'''
    return check_output(cmd, shell=True).decode('utf-8')


tests/test_cli.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from subprocess import check_output, CalledProcessError
import pytest

import pipstat


def test_pipstat_cmd_without_args_exits_with_nonzero_code():
    with pytest.raises(CalledProcessError):
        run_cmd('pipstat')


def test_pipstat_cmd_version():
    assert run_cmd('pipstat -v') == pipstat.__version__ + '\n'
    assert run_cmd('pipstat --version') == pipstat.__version__ + '\n'


def test_no_division_by_zero_in_bargraph():
    assert pipstat.TICK not in pipstat.bargraph({'foo': 0})


def run_cmd(cmd):
    '''Run a shell command `cmd` and return its output.'''
    return check_output(cmd, shell=True).decode('utf-8')


tests/test_package.py
# -*- coding: utf-8 -*-
import os

import responses
import pytest

import pipstat

HERE = os.path.abspath(os.path.dirname(__file__))

def setup_module():
    with open(os.path.join(HERE, 'response.json'), 'r') as fp:
        response_body = fp.read()
    responses.add(
        responses.GET,
        'http://pypi.python.org/pypi/webargs/json',
        body=response_body,
        content_type='application/json'
    )
    responses.add(
        responses.GET,
        'http://pypi.python.org/pypi/nope/json',
        status=404
    )
    responses.start()

def teardown_module():
    responses.stop()

@pytest.fixture
def package():
    return pipstat.Package('webargs')

class TestPackage:

    def test_repr(self, package):
        assert repr(package) == '<Package(name={0!r})>'.format('webargs')

    def test_versions(self, package):
        vers  = ['0.1.0', '0.2.0', '0.3.0', '0.3.1',
                 '0.3.2', '0.3.3', '0.3.4', '0.4.0']
        assert package.versions == vers

    def test_downloads(self, package):
        assert package.downloads == sum(package.version_downloads.values())

    def test_max_version(self, package):
        assert package.max_version == max(package.version_downloads.items(),
                                            key=lambda item: item[1])

    def test_min_version(self, package):
        assert package.min_version == min(package.version_downloads.items(),
                                            key=lambda item: item[1])

    def test_avg_downloads(self, package):
        avg = package.downloads / len(package.versions)
        assert package.average_downloads == int(avg)


pipstat.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""pipstat
Prints download statistics for PyPI packages.

Usage:
    pipstat <package> ...
"""
from __future__ import unicode_literals, print_function, division
import re
import sys
import os
import time
import struct

import math
from collections import OrderedDict

import requests
from dateutil.parser import parse as dateparse

__version__ = "0.3.0-dev"
__author__ = "Steven Loria"
__license__ = "MIT"

DATE_FORMAT = "%y/%m/%d"
MARGIN = 5
TICK = '*'
DEFAULT_PYPI = 'http://pypi.python.org/pypi'
PYPI_RE = re.compile('''^(?:(?P<pypi>https?://[^/]+/pypi)/)?
                        (?P<name>[-A-Za-z0-9.]+)
                        (?:/(?P<version>[-A-Za-z0-9.]+))?$''', re.X)


def lazy_property(fn):
    """Decorator that makes a property lazy-evaluated."""
    attr_name = '_lazy_' + fn.__name__

    @property
    def _lazy_property(self):
        if not hasattr(self, attr_name):
            setattr(self, attr_name, fn(self))
        return getattr(self, attr_name)
    return _lazy_property

def get_terminal_size():
    """Returns the current size of the terminal as tuple in the form
    ``(width, height)`` in columns and rows.

    Credit to Armin Ronacher (BSD Licencsed). See NOTICE file for license info.
    """
    # If shutil has get_terminal_size() (Python 3.3 and later) use that
    if sys.version_info >= (3, 3):
        import shutil
        shutil_get_terminal_size = getattr(shutil, 'get_terminal_size', None)
        if shutil_get_terminal_size:
            sz = shutil_get_terminal_size()
            return sz.columns, sz.lines

    def ioctl_gwinsz(fd):
        try:
            import fcntl
            import termios
            cr = struct.unpack(
                'hh', fcntl.ioctl(fd, termios.TIOCGWINSZ, '1234'))
        except Exception:
            return
        return cr

    cr = ioctl_gwinsz(0) or ioctl_gwinsz(1) or ioctl_gwinsz(2)
    if not cr:
        try:
            fd = os.open(os.ctermid(), os.O_RDONLY)
            try:
                cr = ioctl_gwinsz(fd)
            finally:
                os.close(fd)
        except Exception:
            pass
    if not cr or not cr[0] or not cr[1]:
        cr = (os.environ.get('LINES', 25),
              os.environ.get('COLUMNS', 80))
    return int(cr[1]), int(cr[0])


def bargraph(data):
    """Return a bar graph as a string, given a dictionary of data."""
    lines = []
    max_length = min(max(len(key) for key in data.keys()), 20)
    max_val = max(data.values())
    max_val_length = max(len('{:,}'.format(val)) for val in data.values())
    term_width = get_terminal_size()[0]
    max_bar_width = term_width - MARGIN - (max_length + 3 + max_val_length + 3)
    template = "{key:{key_width}} [ {value:{val_width},d} ] {bar}"
    for key, value in data.items():
        try:
            bar = int(math.ceil(max_bar_width * value / max_val)) * TICK
        except ZeroDivisionError:
            bar = ''
        line = template.format(key=key[:max_length], value=value,
            bar=bar, key_width=max_length, val_width=max_val_length)
        lines.append(line)
    return '\n'.join(lines)


class PackageError(Exception):
    pass


class NotFoundError(PackageError):
    pass


class Package(object):

    def __init__(self, name, client=None, pypi_url=DEFAULT_PYPI):
        self.client = client or requests.Session()
        self.name = name
        self.url = '{pypi_url}/{name}/json'.format(pypi_url=pypi_url, name=name)

    @lazy_property
    def data(self):
        resp = self.client.get(self.url)
        if resp.status_code == 404:
            raise NotFoundError('Package not found')
        return resp.json()

    @lazy_property
    def versions(self):
        """Return a list of versions, sorted by release datae."""
        return [k for k, v in self.release_info]

    @lazy_property
    def version_downloads(self):
        """Return a dictionary of version:download_count pairs."""
        ret = OrderedDict()
        for release, info in self.release_info:
            download_count = sum(file_['downloads'] for file_ in info)
            ret[release] = download_count
        return ret

    @property
    def release_info(self):
        release_info = self.data['releases']
        # filter out any versions that have no releases
        filtered = [(ver, releases) for ver, releases in release_info.items()
                    if len(releases) > 0]
        # sort by first upload date of each release
        return sorted(filtered, key=lambda x: x[1][0]['upload_time'])

    @lazy_property
    def version_dates(self):
        ret = OrderedDict()
        for release, info in self.release_info:
            if info:
                upload_time = dateparse(info[0]['upload_time'])
                ret[release] = upload_time
        return ret

    def chart(self):
        data = OrderedDict()
        for version, dl_count in self.version_downloads.items():
            date = self.version_dates.get(version)
            date_formatted = ''
            if date:
                date_formatted = time.strftime(DATE_FORMAT,
                    self.version_dates[version].timetuple())
            key = "{0:7} {1}".format(version, date_formatted)
            data[key] = dl_count
        return bargraph(data)

    @lazy_property
    def downloads(self):
        """Total download count."""
        return sum(self.version_downloads.values())

    @lazy_property
    def max_version(self):
        """Version with the most downloads."""
        data = self.version_downloads
        return max(data.items(), key=lambda item: item[1])

    @lazy_property
    def min_version(self):
        """Version with the fewest downloads."""
        data = self.version_downloads
        return min(data.items(), key=lambda item: item[1])

    @lazy_property
    def average_downloads(self):
        """Average number of downloads."""
        return int(self.downloads / len(self.versions))

    @property
    def downloads_last_day(self):
        return self.data['info']['downloads']['last_day']

    @property
    def downloads_last_week(self):
        return self.data['info']['downloads']['last_week']

    @property
    def downloads_last_month(self):
        return self.data['info']['downloads']['last_month']

    def __repr__(self):
        return '<Package(name={0!r})>'.format(self.name)


def main():
    """Main entry point for the pipstat CLI. Prints download stats to std out.
    """
    if '-h' in sys.argv or '--help' in sys.argv:
        print(__doc__)
        sys.exit(0)
    if '-v' in sys.argv or '--version' in sys.argv:
        print(__version__)
        sys.exit(0)
    package_names = sys.argv[1:]
    if len(package_names) < 1:
        print("No package names specified.", file=sys.stderr)
        print(__doc__)
        sys.exit(1)
    else:
        client = requests.Session()
        for input_name in package_names:
            m = PYPI_RE.match(input_name)
            if not m:
                print("Invalid name or URL: {name!r}".format(name=input_name),
                      file=sys.stderr)
                continue
            pypi_url = m.group('pypi') or DEFAULT_PYPI
            name = m.group('name')
            url = "{pypi_url}/{name}".format(pypi_url=pypi_url, name=name)
            print("Fetching statistics for '{url}'. . .".format(url=pypi_url))
            package = Package(name, pypi_url=pypi_url, client=client)
            try:
                version_downloads = package.version_downloads
            except NotFoundError:
                print("No versions of {0!r} were found.".format(name), file=sys.stderr)
                sys.exit(1)
            chart = package.chart()
            min_ver, min_downloads = package.min_version
            max_ver, max_downloads = package.max_version
            avg_downloads = package.average_downloads
            total = package.downloads
            print()
            header = "Download statistics for {name}".format(name=name)
            print(header)
            print('=' * len(header))
            print('Downloads by version')
            print(chart)
            print()
            print("Min downloads:   {min_downloads:12,} ({min_ver})".format(**locals()))
            print("Max downloads:   {max_downloads:12,} ({max_ver})".format(**locals()))
            print("Avg downloads:   {avg_downloads:12,}".format(**locals()))
            print("Total downloads: {total:12,}".format(**locals()))
            print()
            print('Last day:    {daily:12,}'.format(daily=package.downloads_last_day))
            print('Last week:   {weekly:12,}'.format(weekly=package.downloads_last_week))
            print('Last month:  {monthly:12,}'.format(monthly=package.downloads_last_month))
    sys.exit(0)

if __name__ == '__main__':
    main()
